crypto/sm2/sm2_pmeth.c

```diff
@@ -212,6 +213,24 @@ static int pkey_sm2_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
         smctx->id_set = 1;
         return 1;
 
+    case EVP_PKEY_CTRL_SET1_HEX_ID:
+        if (p1 > 0) {
+            tmp_id = OPENSSL_hexstr2buf((const char *)p2, &hex_len);
+            if (tmp_id == NULL) {
+                SM2err(SM2_F_PKEY_SM2_CTRL, ERR_R_PASSED_INVALID_ARGUMENT);
+                return 0;
+            }
+            OPENSSL_free(smctx->id);
+            smctx->id = tmp_id;
+        } else {
+            /* set null-ID */
+            OPENSSL_free(smctx->id);
+            smctx->id = NULL;
+        }
+        smctx->id_len = (size_t)hex_len;
+        smctx->id_set = 1;
+        return 1;
+
```

Comment:
In all other parts of OpenSSL where we implement "hex" type string controls, parsing the hex string is done *outside* of the central `ctrl` function...  either in the `ctrl_str` itself, or via an intermediate function.

In this case, it would be quite simple to move the hex parsing to `pkey_sm2_ctrl_str` and call `pkey_sm2_ctrl` with `EVP_PKEY_CTRL_SET1_HEX_ID` and the resulting buffer.

Reply:
(note that keeping the hex parsing to the `ctrl_str` function will help making a transition to OSSL_PARAM style parameter passing)

Reply:
Then `EVP_PKEY_CTRL_SET1_HEX_ID` is not needed anymore

Reply:
Yup.  But since you added it in this PR, that should not be a problem

Reply:
Hmmm, I don't quite understand this. There is direct call to `pkey_sm2_ctrl` by functions like `EVP_PKEY_CTX_ctrl`, so if the hex parsing logic is moved outside of `pkey_sm2_ctrl`, then the hex will not be parsed...

Reply:
That's what `EVP_PKEY_CTX_ctrl_str()` is for.  `EVP_PKEY_CTX_ctrl()` takes raw data.

Reply:
Ahh, so be it.

Reply:
Moved the hex parsing code to `str` function.

I also removed the `EVP_PKEY_CTRL_SET1_HEX_ID` and related macro, since I found it totally not necessary to introduce such a thing... User can just call `EVP_PKEY_CTX_ctrl_str` with specific command strings.